### ISSUE TITLE
sassMiddlewareError never emptied

### ISSUE THREAD

**Issue by jacekkopecky** [sass/node-sass-middleware#93]
When developing, if I make a mistake in my SCSS, node-sass-middleware duly logs it and that's very good. But when I fix the mistake and save the file (and on any further save too) the CSS request that triggers the recompile fails, probably because `sassMiddlewareError` is never emptied.

I'm using it with Express like this:
```
app.use(sassMiddleware({
    src: 'webpages',
    outputStyle: 'expanded',
    response: false,
  }));
```

**Expected behaviour:** an error is logged once, successful recompiles result in successful response.

**Observed behaviour:** once an error happens, every new recompile logs it even if the compile didn't encounter any errors this time; the recompile request gets an internal server error result.

**Comment by am11** [comment]
Hey @jacekkopecky, I quickly tried with [this](https://github.com/sass/node-sass-middleware/issues/70#issuecomment-261822665) example:

```cmd
mkdir c:\temp\projectx\public\sass
cd c:\temp\projectx

echo a{bc;} > c:\temp\projectx\public\sass\test.scss
:: note the error, no colon between b and c

notepad c:\temp\projectx\package.json
:: typed the following package.json
{
  "dependencies": {
    "express": "~4.14.0",
    "node-sass-middleware": "^0.10.1"
  }
}

notepad c:\temp\projectx\my-server.js
:: copy paste from https://github.com/sass/node-sass-middleware/issues/70#issuecomment-261822665

:: finally run the server
node my-server

:: now navigate to
:: http://localhost:8000/static/css/test.css
:: to get the error like `Error: property "bc" must be followed by a ':'`

:: add colon between b and c in test.scss file, save the file and refresh the page
:: to get successful rendered result
```

Please give it a try and let us know if you are still able to reproduce the issue with these steps.

Thanks.

**Comment by jacekkopecky** [comment]
thank you, will do

**Comment by jacekkopecky** [comment]
Hi, yes, the problem persists, like this:

server.js:
```
var sassMiddleware = require('node-sass-middleware');
var app = require('express')();
var serveStatic = require('serve-static');
var join = require('path').join;
var http = require('http');
var port = process.env.PORT || 8000;

var myPrefix = '/static/css';
var destination = join(__dirname, 'public/css');

// Sass middleware
app.use(sassMiddleware({
  /* Options */
  src: join(__dirname, 'public/sass'),
  dest: destination,
  debug: true,
  outputStyle: 'compressed',
  prefix: myPrefix 
}));

// static middleware
app.use(myPrefix, serveStatic(destination));

// start server
http.createServer(app).listen(port);
```

I use `test.scss` in the current directory, just for simplicity, so then I start the server and in Chrome, go to http://localhost:8000/test.css and get the error `Error: property "bc" must be followed by a ':'`

I fix the error by adding a ':' in the right place, in the browser first reload repeats the error, the second reload gives me the right result.

Then I edit test.scss again, keeping it valid; in the browser, the first reload gives me the old error, the second reload gives me the newest state of the CSS file.

The difference from yours is `response: false` - the code has a special case for where `response === false` in one place, and in other places just tests for a falsey value, so `options.response` is actually tri-state in this way. Without `response: false`, the bug doesn't show.

**Comment by am11** [comment]
@jacekkopecky, thanks for taking time looking into the source code. :)

My mistake, in my example above, since we have set the prefix, we should be visiting http://localhost:8000/static/css/test.css instead. The first time it compiles / renders the Sass -> CSS (you can see this trace in console log), all the subsequent requests will skip the compilation until we change the .scss file.

In case you are not using prefix option, you can simplify it by removing the following lines:

1. `var myPrefix = '/static/css';`
2. `prefix: myPrefix`

and static middleware line will look like:

```javascript
app.use(serveStatic(destination));
```

After that invoke the previous URL http://localhost:8000/test.css and refresh the page multiple times.

**Comment by jacekkopecky** [comment]
sorry, I had to edit my comment above - I pasted the wrong code! Now it shows my configuration.

**Comment by am11** [comment]
@jacekkopecky, I have reproduced a bit different issue:

1. If there is test.css, delete it before running the server.
2. Before running the server, have the error in CSS file.
3. After running the server navigate to `localhost:8080/test.css`.
4. Correct, the error, save the file and refresh the page while the server was still running. At this point it still gives the error. Refresh the page once again and the error disappears no matter how many times we refresh the page.

Is this the same issue as yours? i.e. just after correcting the issue the next refresh does not succeed but all the subsequent refreshes are fine.

With the following patch, I have fixed the issue:

```diff
diff --git a/middleware.js b/middleware.js
index f5e1d87..6a4a19e 100644
--- a/middleware.js
+++ b/middleware.js
@@ -61,7 +61,7 @@ module.exports = function(options) {
     options = { src: options };
   }

-  var sassMiddlewareError = null;
+  var sassMiddlewareError;
   var cachedErrorCb = options.error;

   // This function will be called if something goes wrong
@@ -109,6 +109,8 @@ module.exports = function(options) {

   // Middleware
   return function sass(req, res, next) {
+    sassMiddlewareError = null;
+
     if (req.method != 'GET' && req.method != 'HEAD') {
       return next();
     }
```

(reset the middleware error object per compile)

Could you please give it a try?

**Comment by jacekkopecky** [comment]
Hi, yes, this quick fix works.

Is there an issue with asynchrony, though? The middleware could be running several times concurrently (as the various FS and compilation tasks interleave with incoming HTTP requests) so the `sassMiddlewareError` variable shouldn't be global for the whole middleware, rather it should be local for the function `sass`. Should the variable be defined on line 112, and should the `error` function also be therefore defined inside the `sass` function?

**Comment by am11** [comment]
Thanks for the confirmation @jacekkopecky! 👍 

I agree, we can refactor it a bit and provide a decent asynchronous middleware with no global variables.

Would you like to send a pull request? Otherwise, it will make these fixes over the weekend.

**Comment by jacekkopecky** [comment]
I'll prepare a PR @am11

**Comment by am11** [comment]
Fixed by #95 👌
